File: 알고리즘 스터디 문제들/Part 5 - 다이나믹 프로그래밍/추가 문제/퇴사.cpp
#include <iostream>
#include <array>
using namespace std;

// con[i].first == T[i + 1]
// con[i].second = P[i + 1]
array<pair<int, int>, 16> con;

// dp[i] == (i + 1)번째 날의 상담을 시작하기 전에 얻을 수 있는 최대 이익
array<int, 16> dp = {};

int main()
{
    size_t N;
    cin >> N;
    for (size_t i = 0; i < N; i++)
        cin >> con[i].first >> con[i].second;

    int ans = 0;
    for (size_t i = 0; i < N; i++)
    {
        for (size_t j = i + 1; j <= N; j++)
        {
            if (i + con[i].first <= j && dp[i] + con[i].second > dp[j])
            {
                dp[j] = dp[i] + con[i].second;
                ans = max(ans, dp[j]);
            }
        }
    }
    cout << ans;
}

File: 알고리즘 스터디 문제들/Part 5 - 다이나믹 프로그래밍/추가 문제/퇴사.py
N = int(input())
con = [[0, 0] for i in range(N + 1)]
dp = [0 for i in range(N + 1)]
for i in range(N):
    con[i][0], con[i][1] = map(int, input().split())

for i in range(1, N + 1):
    for j in range(i):
        if (dp[i] < dp[j] + con[j][1] and i >= j + con[j][0]):
            dp[i] = dp[j] + con[j][1]

print(max(dp))
